### PR TITLE
New contributing.md

### PR DESCRIPTION
**Pull request by RedXanadu**
PR brings the CONTRIBUTING.MD document in line with the updated, HTML-formatted contribution guidelines from the `/docs` site. Also adds a hyperlink to said HTML content.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,159 +1,307 @@
...
+    * Bonus points for submitting tests along with the issue.
+3. Fork the repository on GitHub and begin making changes there.
+4. Signed commits are preferred. (For more information and help with this, refer to the [GitHub documentation](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)).
 
 ## Making Changes
 
-* Please base your changes on the latest dev branch (e.g. ```v3.4/dev```)
-* Create a topic branch for your feature or bug fix.
-* Please fix only one problem at a time; this will help to quickly test and merge your change. If you intend to fix multiple unrelated problems, please use a separate branch for each problem.
+* Base any changes on the latest dev branch (e.g., `v3.4/dev`).
```
[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md:19]: The idea was to have this file auto-generated from the online version, but I don't know how to do that… (maybe a GitHub action or something?)

I'll remove the explicit version number for the time being.

[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md]: Removed.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,159 +1,307 @@
...
-* Make sure your commit messages are in the [proper format](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html): The first line of the message should have 50 characters or less, separated by a blank line from the (optional) body. The body should be wrapped at 70 characters and paragraphs separated by blank lines. Bulleted lists are also fine.
+* Make sure commits adhere to the contribution guidelines presented in this document.
+* Make sure commit messages follow the [standard Git format](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
+
+## General Formatting Guidelines for Rules Contributions
+
+* American English should be used throughout.
+* 4 spaces should be used for indentation (no tabs).
+* No trailing whitespace at EOL.
+* No trailing blank lines at EOF.
```
[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md:31]: Ah, is that a new thing with the new linter? I think it's said "no trailing lines at EOF" for years.

So, replace with "Must have a trailing line at EOF"?

@fzipi Git blame says you added that line in originally :P Does this sound right?

File: CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (lifeforms) [reply CONTRIBUTING.md]: We have a release procedure https://github.com/coreruleset/coreruleset/wiki/Release-Procedure and updating version numbers all over the place is part of that, it could be added there.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (theseion) [reply CONTRIBUTING.md:31]: I just checked the linter configuration and it requires a newline as the last character before EOF.

File: CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md:19]: Good thinking!

Branch version now added back in.
New instruction added to the release procedure document.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md:31]: Thanks for checking. So, in that case, the instructions are correct as they currently are, right?

[the remark this answers is not in the document]
Comment (theseion) [reply CONTRIBUTING.md:31]: Maybe it's just the wording but "no trailing blank lines" to me means that the last character at EOF must _not_ be a newline.

The last point makes the second to last obsolete, doesn't it? I would keep the last point and remove the second to last.

[the remark this answers is not in the document]
Comment (fzipi) [reply CONTRIBUTING.md:31]: Hmmm.... "no trailing blank lines" is supposed to cover multiple empty lines at EOF. If we remove it, the other two won't cover that one, right? We need it.

**Comment by fzipi** [comment]
Looking good! I would add a paragraph about testing maybe? Should cover:
- adding positive tests
- adding negative tests
- be consistent using headers (do we have a template for this one? 🤔 )
- when to use encoded_request (and deprecate using raw_request)
Sounds reasonable?

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (theseion) [reply CONTRIBUTING.md:31]: Right. "Files must end in a single newline character". That should do it.

**Comment by fzipi** [comment]
@RedXanadu Hey! Do you need help here?

**Comment by RedXanadu** [comment]
@fzipi I'm not 100% clear about the 'writing tests' part.

"adding positive tests, adding negative tests"
Definitely, two bullet points to add. This part makes sense.

"be consistent using headers (do we have a template for this one?)"
I'm not sure about this. I know we have `positivetest.yaml.skeleton` but I definitely think that's more confusing than helpful… Do we have a simple example test to document?

"when to use encoded_request (and deprecate using raw_request)"
I'm not sure what this means.

**Comment by RedXanadu** [comment]
How about using these as example tests, as I think they're relatively easy to understand:

## Positive Test Example

```
 - test_title: 932100-21
    desc: "Unix command injection"
    stages:
      - stage:
          input:
            dest_addr: 127.0.0.1
            headers:
              Host: localhost
              User-Agent: OWASP ModSecurity Core Rule Set
              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
            method: POST
            port: 80
            uri: "/"
            data: "var=` /bin/cat /etc/passwd`"
            version: HTTP/1.0
          output:
            log_contains: id "932100"
```

## Negative Test Example

```
 - test_title: 932150-5
    stages:
      - stage:
          input:
            dest_addr: "127.0.0.1"
            method: "POST"
            port: 80
            headers:
              User-Agent: OWASP ModSecurity Core Rule Set
              Host: "localhost"
              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
            data: 'foo=ping pong tables'
            protocol: "http"
          output:
            no_log_contains: "id \"932150\""
```

**Comment by fzipi** [comment]
One of the things I've realised long before doing the change is... we should use `Accept: */*`. Shorter. Easier. 🤷 

Other than that, I think the description (`desc` field) is also important. Because that is where you refer to why this is negative test.

Regarding `encoded_request` and `raw_request`, those are possible ways of encoding requests. They allow you to encapsulate everything, so you don't declar headers for example. The test will send exactly what is inside that field. But using `raw` is difficult to maintain. You might be using weird bytes that might break yaml or parsers. So we should recommend to use `encoded_request` only, because it will do exactly the same, but it more portable.

**Comment by RedXanadu** [comment]
Okay, great, that makes sense. Thanks.

I'll take a look at writing this up tomorrow.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (RedXanadu) [reply CONTRIBUTING.md:31]: Great. Unless there are any objections, I'm going with the following:

- Files must end with a single newline character.
- No trailing whitespace at EOL.
- No trailing blank lines at EOF (only the required single EOF newline character is allowed).

I added the explanation in brackets because otherwise "Files must end with a single newline character" and "No trailing blank lines at EOF" seem contradictory.

[the remark this answers is not in the document]
Comment (theseion) [reply CONTRIBUTING.md:31]: Perfect.

**Comment by RedXanadu** [comment]
Content agreed and merged in the documentation repo.

Now updating this PR to include the agreed-upon changes.
